### PR TITLE
[IDE] migrate `test` folder to CMake configuration

### PR DESCRIPTION
**Pull request by Arpafaucon**
As discussed via discord, here's a proposed PR to migrate test. 

Some notes
- I kept `DEBUGINFOD` enabled by default on linux platforms like in the `MakeFile`
- we "include" the top-level `CmakeLists.txt` to get the `TracyClient` target - that's a bit unorthodox but seems to work well
- on my Ubuntu machine, I don't need to link explicitly against `pthread` but IIRC some other Linux need it, so I kept it explicit for everyone
- `-ldl` link option seems implicitly added by CMake, so I don't mention it explictly
- updated `std=` option from `gnu++11` to `gnu++20` to align with other projects
- due to the way `CMake` works, the `-Wall -g3 -fmerge-constants` options are not propagated anymore to `TracyClient.cpp` (because it's a different library) - people will have to add them explicitly for debugging if needed. 

For comparison, here are the new compilation lines (I had disabled DEBUGINFOD for this run)

```
/usr/bin/c++ -DTRACY_ENABLE -isystem /home/gregoire.roussel/dev/perf/tracy/public -std=gnu++20 -MD -MT client/CMakeFiles/TracyClient.dir/public/TracyClient.cpp.o -MF CMakeFiles/TracyClient.dir/public/TracyClient.cpp.o.d -o CMakeFiles/TracyClient.dir/public/TracyClient.cpp.o -c /home/gregoire.roussel/dev/perf/tracy/public/TracyClient.cpp
/usr/bin/c++ -DTRACY_ENABLE -isystem /home/gregoire.roussel/dev/perf/tracy/public -Wall -g3 -fmerge-constants -std=gnu++20 -MD -MT CMakeFiles/tracy-test.dir/test.cpp.o -MF CMakeFiles/tracy-test.dir/test.cpp.o.d -o CMakeFiles/tracy-test.dir/test.cpp.o -c /home/gregoire.roussel/dev/perf/tracy/test/test.cpp
/usr/bin/c++ -rdynamic CMakeFiles/tracy-test.dir/test.cpp.o -o tracy-test  client/libTracyClient.a -lpthread -ldl
```

and the compilation lines of the makefile on master
```
g++ -c -I../public/tracy -g3 -fmerge-constants -Wall -DTRACY_ENABLE -std=gnu++11  test.cpp -o test.o
g++ -c -I../public/tracy -g3 -fmerge-constants -Wall -DTRACY_ENABLE -std=gnu++11  ../public/TracyClient.cpp -o ../public/TracyClient.o
g++ -g3 -fmerge-constants -Wall -DTRACY_ENABLE -std=gnu++11  test.o ../public/TracyClient.o -lpthread -ldl -rdynamic -o tracy_test
```

## Additional changes

I also added a safeguard around the use of the image to detect when the loading failed. It took me some time to understand I was not running the executable from the right location; I hope error lines will help the next dev :)

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
+cmake_minimum_required(VERSION 3.16)
+
+option(USE_DEBUGINFOD "Compile with debuginfod integration" ON)
+
+include(${CMAKE_CURRENT_LIST_DIR}/../cmake/version.cmake)
+
+set(CMAKE_CXX_STANDARD 20)
```
Comment (wolfpld) [inline test/CMakeLists.txt:7]: The client part should target C++11.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  VERSION ${TRACY_VERSION_STRING})
+
+# a bit weird but works: include the client cmake config coming from top-level
+# cmake needs us to specify the build subfolder -> client/ this way we can
+# simply link the test executable against TracyClient
+add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/.. client/)
+
+add_executable(tracy-test test.cpp)
+target_link_libraries(tracy-test TracyClient pthread)
+target_compile_options(tracy-test PRIVATE -Wall -g3 -fmerge-constants)
```
Comment (wolfpld) [inline test/CMakeLists.txt:21]: It's ok to drop this.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  LANGUAGES C CXX
+  VERSION ${TRACY_VERSION_STRING})
+
+# a bit weird but works: include the client cmake config coming from top-level
+# cmake needs us to specify the build subfolder -> client/ this way we can
+# simply link the test executable against TracyClient
+add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/.. client/)
+
+add_executable(tracy-test test.cpp)
+target_link_libraries(tracy-test TracyClient pthread)
```
Comment (wolfpld) [inline test/CMakeLists.txt:20]: Let's go without `pthread` if it works for you.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+target_compile_options(tracy-test PRIVATE -Wall -g3 -fmerge-constants)
+target_link_options(tracy-test PRIVATE -rdynamic)
+
+# OS-specific options
+exec_program(uname ARGS -s
+  OUTPUT_VARIABLE SYSTEM_KERNEL)
+exec_program(uname ARGS -o
+  OUTPUT_VARIABLE SYSTEM_OS)
```
Comment (wolfpld) [inline test/CMakeLists.txt:28]: CMake specific platform variables should be used instead (`CMAKE_SYSTEM_NAME`).

File: test/test.cpp [hunk 5]
```diff
@@ -328,7 +328,10 @@ int main()
 
     int x, y;
     auto image = stbi_load( "image.jpg", &x, &y, nullptr, 4 );
-
+    if(image == nullptr)
+    {
+        std::cerr << "Could not find image (are you running this executable in the test/ folder ?), skipping" << std::endl;
```
Comment (wolfpld) [inline test/test.cpp:333]: This seems like a misconfiguration on the CMake part. Can you make the image file be copied to the build / install dir instead?

**Comment by Arpafaucon** [comment]
Like always, I forgot about the CI configuration. I updated it for CMake, and I had to add the advanced option `TRACY_DEMANGLE` to match the existing test

**Pull request merged**